server/auth/auth.controller.js

```diff
@@ -0,0 +1,184 @@
+const bcrypt = require('bcrypt');
+
+const logger = require('../../config/logger')(module);
+const joiErrMsg = require('../utils/JoiErrorMessage');
+const asyncHandler = require('../middleware/async');
+const ErrRes = require('../utils/errorResponse');
+const jwt = require('jsonwebtoken');
+const config = require('../../config/config');
+
+const redis = require('../../index.redis');
+const tokenGenerator = require('../utils/tokenGenerator');
+
+const {
+  userValidation,
+  adminValidation,
+  loginValidation,
+  verifyAccountValidation
+} = require('./auth.validation');
+const { userModel, adminModel } = require('./auth.model');
+const { sendMailForVerifyAccount } = require('../../config/sendMail');
+
+exports.registerUser = asyncHandler(async (req, res, next) => {
+  const { error } = userValidation(req.body);
+  if (error) return next(new ErrRes(joiErrMsg(error), 400));
+
+  const {
+    fullname,
+    email,
+    password,
+    confirmPassword,
+    location,
+    contactNo,
+    role,
+    facebookID,
+    gmailID
+  } = req.body;
+
+  if (password !== confirmPassword)
+    return next(new ErrRes('Password is not matching', 401));
+
+  if (facebookID) {
+    const isPresent = await userModel.findOne({ facebookID });
+    if (isPresent)
+      return res.status(400).json({
+        success: false,
+        message: 'Account is already created with this ID'
+      });
+  }
+  if (gmailID) {
+    const isPresent = await userModel.findOne({ gmailID });
+    if (isPresent)
+      return res.status(400).json({
+        success: false,
+        message: 'Account is already created with this ID'
+      });
+  }
+
+  const hashedPassword = await bcrypt.hash(password, 10);
+  const user = new userModel({
+    fullname,
+    email,
+    hashedPassword,
+    location,
+    contactNo,
+    role,
+    facebookID,
+    gmailID
+  });
+  await user.save();
+  const verifyToken = tokenGenerator();
+  await sendMailForVerifyAccount(fullname, email, verifyToken, 'USER');
```

Comment:
Explain this line

Reply:
as I said before, I am sending two keywords. 'USER' and 'ADMIN'. It is used here for less query to search.